exploration/expression-attributes.md

```diff
@@ -204,38 +217,125 @@ the reserved/private-use rules will need to be adjusted to support attributes.
+
+Additional restrictions could be imposed,
+e.g. requiring that each `u:id` is unique within a formatted message.
+
+### Attributes
+
+Add support for standalone `@key` as well as
+option-like `@key=value` attribute pairs with a literal value
+at the end of any expression or markup.
+
+To distinguish attributes from options,
 require `@` as a prefix for each attribute asignment.
-Examples: `@translate=yes` and `@locale=$exprLocale`.
+Examples: `@translate=yes` and `@xliff:canCopy`.
 
-Define the meaning and supported values of some expression attributes in the specification,
-including at least `@dir` and `@locale`.
+Do not allow expression or markup attributes to influence the formatting context,
+or pass them to function handlers.
 
-To support later extension of the specified set of attributes while allowing user extensibility,
-suggest custom attribute names to include a U+002D Hyphen-Minus `-`.
-Examples: `@can-copy=no`, `@note-link=|https://...|`.
+Drop variable values from the `attribute` rule:
 
-Allow expression attributes to influence the formatting context,
-but do not directly pass them to user-defined functions.
+```diff
+-attribute = "@" identifier [[s] "=" [s] (literal / variable)]
++attribute = "@" identifier [[s] "=" [s] literal]
+```
 
 ## Alternatives Considered
 
-### Do not support expression attributes
+### Do nothing
+
+Continue to [caution](https://github.com/unicode-org/message-format-wg/blob/d38ff326d2381b3ef361e996c3431d1b251518d6/spec/syntax.md#attributes)
+function authors and other implementers away from creating function-specific or implementation-specific option values
+for the use cases presented above.
+
+As should be obvious, the current situation is not tenable in the long term, and should be resolved.
+
+### Do not provide any guidance
+
+Do not include in the spec rules or guidance for declaring formatted part identifiers,
+or overriding the message locale or directionality.
+
+Do not define a common way to communicate information
+about an expression or markup to translators or tools.
+
+This would mean not defining anything for default registry functions either,
+effectively requiring implementation-specific options like `icu:locale`.
+
+Other functions could use their own definitions and handling for similar options,
+such as `locale` or `x:lang`.
+
+Formatted parts for markup would not be able to directly include an identifier.
 
 If not explicitly defined, less information will be provided to translators.
 
 Function options may be used as a workaround,
 but each implementation and user will end up with different practices.
 
-### Use function options, but with some suggested prefix like `_`
+### Define options for default registry only
 
-A bit less bad than the previous, but still mixes attributes and options into the same namespace.
+Define at least `locale` and `dir` as options for default registry functions,
+with handling internal to each function implementation.
```

Comment:
I think you should address how `id` _could_ work here. Presumably the `id` property of the resolved value could be queried?

Reply:
I'm honestly not sure how to make `id` sensibly work, if we're not defining it generally in the spec but only specifically for each standard registry function. That would mean not having any specific field for it in the formatting-internal resolved values, and intermingling it with the other resolved options.

I hope we do not end up with this option, and so can avoid needing to answer all the questions that it would introduce.

Reply:
I agree that we should not end up with this option. I'd probably address it by saying:

```suggestion
Define the `locale`, `dir`, and `id` options for all default registry functions
and establish a policy of including them in all future default registry functions,
with the hope that other functions and implementations will adopt this as a best practice.
Actually implementation would be internal to each function.
```

Reply:
If we were to define `id` for default registry functions, we would also need to define what it does for them. Could you give some example for how you think that could be done without also defining `id` as somehow special in the spec?

Reply:
Oh no, I agree with you there. These would be somehow special. If we don't reserve them outright for all functions everywhere (and instead do it by policy and hope for imitation by local functions), then people will be confused later when local functions work funny. Note that this is one argument for why expression annotations make sense to me: since they are not options, there is no danger of conflict or redefinition.

Reply:
Given the clarifying text about hacky identifiers added a few lines below this, is there anything actionable left in this thread?